### DISCUSSION TITLE
Non typed filter variable will mess up query return types

### DISCUSSION THREAD

**Discussion by tudor33sud** [mikro-orm/mikro-orm#5060] [Q&A] [answered]
Using $or operator to filter entities generates correct query and correct results, however the type will get messed up if we create a filter variable object with no type and reference that in find methods. 

Behaviour: the query result assumes that the returned objects will not be of the entity class itself, but only the subset of properties that were present within the filters array. Please check the comments in the repro to get a better idea of what kind of issues arise. You will see that return types will be Loaded<those proops> instead of Loaded<UserEntity>

I'm not sure whether this is just a Typescript Limitation that we should overcome by typing filters, or if there's something better that can be added in MikroORM types definitions. I tried to have a look at the underlying types but it's quite overwhelming to go through all of them, maybe  someone from the team will know faster if this is really a MikroORM bug or not.

PS: node_modules were not included, please npm install in order to see the types.

[types-reproduction.zip](https://github.com/mikro-orm/mikro-orm/files/13793704/types-reproduction.zip)

**Comment by B4nan** [comment, marked as answer]
Yeah that's a TS limitation, I am not aware of any way to enforce inference only for the first parameter. I know there is one trick to disallow inference on some parameters but it didn't work in this case (and resulted in too complex types and cycle issues). There was some recent movement but afaik it was postponed, TS should at some point allow partial type inference which should help here too I hope.

There are several ways around this:

1. type the query explicitly

```ts
const nonTypedFilter = {
    $or: [
      { email: 'Foo', name: 'foo' },
      { email: 'Bar', name: 'bar' },
    ],
} as ObjectQuery<User>;
const noLongerbadTypedUser = await orm.em.findOne(User, nonTypedFilter);
// Loaded<User, never, "*", never> | null
```

2. use a repository which ensures correct entity type
```ts
const nonTypedFilter = {
    $or: [
      { email: 'Foo', name: 'foo' },
      { email: 'Bar', name: 'bar' },
    ],
};
const anotherNoLongerbadTypedUser = await orm.em.repo(User).findOne(nonTypedFilter);
// Loaded<User, never, "*", never> | null
```

**Reply by tudor33sud** [reply]
Thanks a lot for your fast answer! I was kind of suspecting, but I was curious about which types do you think we should use in these referenced filters scenario. 

Happy Holidays for you! We are having a blast using MikroORM, your effort is certainly seen in the community :)
